src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -596,26 +596,36 @@ impl<const N: usize> Inner<N> {
         }
     }
 
+    /// Processes an incoming TCP segment.
     fn receive(&self, ip_hdr: &Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
-        let (mut tcp_hdr, data) = TcpHeader::parse(ip_hdr, buf, self.tcp_config.get_rx_checksum_offload())?;
+        let (mut tcp_hdr, data): (TcpHeader, DemiBuffer) =
+            TcpHeader::parse(ip_hdr, buf, self.tcp_config.get_rx_checksum_offload())?;
         debug!("TCP received {:?}", tcp_hdr);
-        let local = SocketAddrV4::new(ip_hdr.get_dest_addr(), tcp_hdr.dst_port);
-        let remote = SocketAddrV4::new(ip_hdr.get_src_addr(), tcp_hdr.src_port);
+        let local: SocketAddrV4 = SocketAddrV4::new(ip_hdr.get_dest_addr(), tcp_hdr.dst_port);
+        let remote: SocketAddrV4 = SocketAddrV4::new(ip_hdr.get_src_addr(), tcp_hdr.src_port);
 
         if remote.ip().is_broadcast() || remote.ip().is_multicast() || remote.ip().is_unspecified() {
-            return Err(Fail::new(libc::EINVAL, "invalid address type"));
+            let cause: String = format!("invalid remote address (remote={})", remote.ip());
+            error!("receive(): {}", &cause);
+            return Err(Fail::new(libc::EBADMSG, &cause));
         }
 
-        // grab the queue descriptor based on the incoming.
+        // Retrieve the queue descriptor based on the incoming segment.
         let &qd: &QDesc = match self.addresses.get(&SocketId::Active(local, remote)) {
             Some(qdesc) => qdesc,
             None => match self.addresses.get(&SocketId::Passive(local)) {
                 Some(qdesc) => qdesc,
-                None => return Err(Fail::new(libc::EBADF, "Socket not bound")),
+                None => {
```

Comment:
Should we also send a reset here? It seems like this is a similar case to inactive or closed?

Reply:
I am not sure which case you are talking about.

Reply:
Not really. This is the case where we have received a segment for a connection that is not established or not in its way to be established. We just ignore the packet and remote will re-transmit if that is a packet that arrived before the SYN.